ntoskrnl/mm/balance.c

```diff
@@ -39,6 +39,18 @@ static LONG PageOutThreadActive;
 
 /* FUNCTIONS ****************************************************************/
 
+static inline VOID UpdatePeakCommitment(VOID)
+{
+    // HACK: MmTotalCommittedPages should be adjusted consistently with
+    // other counters at different places.
+    MmTotalCommittedPages = MiMemoryConsumers[MC_SYSTEM].PagesUsed +
```

Comment:
Does the caller of this function guarantee that some kind of global lock is held? If yes, please document that (or assert in code that the lock is held if possible, as well as use SAL to annotate the function).
If there's no lock, this should likely be an `InterlockedCompareExchange` loop. Depending on how critical the accuracy of this value is (is it used to enforce memory limits or only for returning information to user mode callers?), the loop might have to repeat until it observes that none of the three source values have changed during the calculation.

Reply:
Actually, given how the `PagesUsed` and `MiUsedSwapPages` counters are updated, perhaps this function should just take a parameter to indicate increment/decrement, and do an `InterlockedIncrement` or `InterlockedDecrement` on the Total value. Much simpler than turning this into a loop or adding a lock.

Reply:
the "lock" discussion goes beyond my understanding of this module.
The current criticity is very low (display of "peak" data in TaskMgr), so i'd propose to go on with this "first simple" implementation and let another more experienced dev continue with another PR to make it safer and reader for more critical use case, if any. Would it make sense to you @ThFabba & @HBelusca ? Otherwise, i'd need more explicit contributions from you to this PR on this specific item :-)

Reply:
You can just use my second suggestion, it's much simpler. Something like this should do the trick -- and then you call it with either +1 or -1 depending on whether the calling code just increased or decreased the commitment.
```
inline VOID UpdateTotalCommittedPages(LONG Increase)
{
    SIZE_T TotalCommittedPages = InterlockedExchangeAddSizeT(&MmTotalCommittedPages, Increase) + Increase;

    SIZE_T PeakCommitment = MmPeakCommitment;
    while (TotalCommittedPages > PeakCommitment &&
           InterlockedCompareExchangeSizeT(&MmPeakCommitment, TotalCommittedPages, PeakCommitment) != PeakCommitment)
    {
        PeakCommitment = MmPeakCommitment;
    }
}
```

Reply:
Yes, please use @ThFabba suggestion. At first I thought this was done under PFN lock, but that's obviously not the case.

Reply:
> LONG Increase

Good point. To further optimize, we could have 2 functions, as decrease never updates peak.

Reply:
That's a pointless optimization. Just doing `if (Increase < 0) return;` before the while() loop should be sufficient.

Reply:
My suggestion is to reduce compiled code size and increase execution speed. (And I find it more explicit / simpler.)
As always, do it as you prefer..

Reply:
> to reduce compiled code size and increase execution speed

So we understand you actually did all those verifications before proposing your suggestions?

Reply:
Implementation done. I'm testing and i'll commit after if works fine.

Reply:
> you actually did all those verifications

No, I simply assume that less parameters/variables/code in C results in less binary code.

Reply:
Are you forgetting that you'd also be duplicating the function prologue/epilogue, alignment and the InterlockedExchangeAdd? It's not at all obvious that it would result in smaller code size - and it definitely wouldn't result in less C code, so I'm not sure how you get to this conclusion.